boa/src/syntax/parser/cursor/mod.rs

```diff
@@ -0,0 +1,166 @@
+//! Cursor implementation for the parser.
+mod buffered_lexer;
+
+use super::ParseError;
+use crate::syntax::{
+    ast::Punctuator,
+    lexer::{InputElement, Lexer, Position, Token, TokenKind},
+};
+use buffered_lexer::BufferedLexer;
+use std::io::Read;
+
+/// The result of a peek for a semicolon.
+#[derive(Debug)]
+pub(super) enum SemicolonResult<'s> {
+    Found(Option<&'s Token>),
+    NotFound(&'s Token),
+}
+
+/// Token cursor.
+///
+/// This internal structure gives basic testable operations to the parser.
+#[derive(Debug)]
+pub(super) struct Cursor<R> {
+    buffered_lexer: BufferedLexer<R>,
+}
+
+impl<R> Cursor<R>
+where
+    R: Read,
+{
+    /// Creates a new cursor with the given reader.
+    #[inline]
+    pub(super) fn new(reader: R) -> Self {
+        Self {
+            buffered_lexer: Lexer::new(reader).into(),
+        }
+    }
+
+    #[inline]
+    pub(super) fn set_goal(&mut self, elm: InputElement) {
+        self.buffered_lexer.set_goal(elm)
+    }
+
+    #[inline]
+    pub(super) fn lex_regex(&mut self, start: Position) -> Result<Token, ParseError> {
+        self.buffered_lexer.lex_regex(start)
+    }
+
+    #[inline]
+    pub(super) fn next(&mut self) -> Result<Option<Token>, ParseError> {
+        self.buffered_lexer.next(true)
+    }
+
+    #[inline]
+    pub(super) fn peek(&mut self, skip_n: usize) -> Result<Option<&Token>, ParseError> {
```

Comment:
These need documentation. Actually, it makes more sense to have it here than on the buffer, since this is the public interface. I wouldn't remove it from there, though.